File: Mushroom/Cells/Dirt.cs
using System;
using System.Collections.Generic;
using Godot;
using Mushroom.Data;
using Mushroom.Mushroom.Data;

namespace Mushroom.Ceils;

[Spawnable]
public class Dirt : CellBase
{
    private const float MaxDampness = 1.0f;
    private const float MinDampness = 0.0f;
    private const float EvaporationRate = 0.002f; 
    private const float GravityFlowRate = 0.05f; 
    private const float HorizontalFlowRate = 0.02f; 
    private const float RetentionCapacity = 0.45f; 
    private const float NutrientFlowRate = 0.005f;

    private static readonly Color DryColor = new Color(0.63f, 0.5f, 0.38f); 
    private static readonly Color WetColor = new Color(0.22f, 0.15f, 0.06f);

    private static readonly Color GrassColor = new Color(0.27f, 0.55f, 0.25f);
    private static readonly Color DeadGrassColor = new Color(0.55f, 0.51f, 0.25f);
    private static readonly float GrassWaterConsumption = 0.00001f;
    private static readonly int DeadGrassTicks = 100;

    public float Dampness { get; set; } = 0.9f;
    public float Nutrients { get; set; } = 1.0f;

    private int _ticksWithoutWater = 0;
    private bool _isGrass = false;
    
    public override Action? Do(Vector2I vector2)
    {
        float nextDampness = Dampness;
        float nextNutrients = Nutrients;
        bool nextIsGrass = _isGrass;
        
        var pendingChanges = new List<(Dirt cell, float dampnessDelta, float nutrientDelta)>();
        
        float currentDampnessChange = 0;
        float currentNutrientChange = 0;
        
        float excessWater = Math.Max(0, nextDampness - RetentionCapacity);
        
        var directions = new[] { Direction.Up, Direction.Down, Direction.Left, Direction.Right };
        
        foreach (var dir in directions)
        {
            var neighborCell = Grid.GetNeighbor(vector2, dir);
            
            if (dir == Direction.Up)
            {
                if (neighborCell is Air && nextDampness > MinDampness)
                {
                    nextIsGrass = true;

                    float evapMulti = Dampness > RetentionCapacity ? 1.0f : 0.2f;
                    currentDampnessChange -= EvaporationRate * evapMulti;

                    if (Random.Shared.NextDouble() < 0.001)
                        currentNutrientChange += 0.01f;
                }
                else if(neighborCell is not Air)
                    nextIsGrass = false;
            }
            
            if (neighborCell is Dirt dirtNeighbor)
            {
                float dampFlow = 0;
                float nutrFlow = 0;
                
                if (nextDampness > dirtNeighbor.Dampness)
                {
                    if (dir == Direction.Down)
                    {
                        float desiredFlow = (nextDampness - dirtNeighbor.Dampness) / 2 * GravityFlowRate;
                        dampFlow = excessWater > 0 ? Math.Min(excessWater, desiredFlow) : desiredFlow / 10f;
                    }
                    else if ((dir == Direction.Left || dir == Direction.Right) && excessWater > 0)
                        dampFlow = Math.Min(excessWater, (nextDampness - dirtNeighbor.Dampness) / 2 * HorizontalFlowRate);
                }
                
                if (nextNutrients > dirtNeighbor.Nutrients)
                {
                    nutrFlow = (nextNutrients - dirtNeighbor.Nutrients) / 2 * NutrientFlowRate * Math.Max(0.1f, nextDampness);
                }

                if (dampFlow > 0 || nutrFlow > 0)
                {
                    currentDampnessChange -= dampFlow;
                    currentNutrientChange -= nutrFlow;
                    pendingChanges.Add((dirtNeighbor, dampFlow, nutrFlow));
                }
            }
        }
        
        if (Math.Abs(currentDampnessChange) > 0.0001f || Math.Abs(currentNutrientChange) > 0.0001f || pendingChanges.Count > 0)
        {
            return () =>
            {
                _isGrass = nextIsGrass;
                
                Dampness = Math.Clamp(Dampness + currentDampnessChange, MinDampness, MaxDampness);
                Nutrients = Math.Clamp(Nutrients + currentNutrientChange, 0f, 1f);
                
                foreach (var change in pendingChanges)
                {
                    change.cell.Dampness = Math.Clamp(change.cell.Dampness + change.dampnessDelta, MinDampness, MaxDampness);
                    change.cell.Nutrients = Math.Clamp(change.cell.Nutrients + change.nutrientDelta, 0f, 1f);
                }

                if (_isGrass && Grid.GetNeighbor(vector2, Direction.Down) is Dirt dirt && dirt.Dampness > MinDampness)
                {
                    dirt.Dampness -= GrassWaterConsumption;
                    _ticksWithoutWater--;
                }
                else
                    _ticksWithoutWater++;
            };
        }
        
        return () =>
        {
            _isGrass = nextIsGrass;
            
            if (_isGrass && Grid.GetNeighbor(vector2, Direction.Down) is Dirt dirt && dirt.Dampness > MinDampness)
            {
                dirt.Dampness -= GrassWaterConsumption;
                _ticksWithoutWater--;
            }
            else
                _ticksWithoutWater++;
        };
    }

    public override Color GetColor(Vector2I vector2)
    {
        if (_isGrass)
        {
            float k = Math.Clamp(_ticksWithoutWater / (float)DeadGrassTicks, 0f, 1f);
            return GrassColor.Lerp(WetColor, k);
        }
        float t = Math.Clamp(Dampness / MaxDampness, 0f, 1f);
        return DryColor.Lerp(WetColor, t);
    }

    public override Color GetUiColor()
    {
        return DryColor.Lerp(WetColor, 0.5f);
    }
}

File: Core/src/cells/dirt.rs
use crate::cells;
use crate::cells::{Action, Cell, AsAny};
use crate::data::{Direction, Position};
use crate::grid::grid;
use crate::cells::air::Air;
use std::sync::{Arc, Mutex};

const MAX_DAMPNESS: f32 = 1.0;
const MIN_DAMPNESS: f32 = 0.0;
const EVAPORATION_RATE: f32 = 0.005;
const GRAVITY_FLOW_RATE: f32 = 0.2;
const HORIZONTAL_FLOW_RATE: f32 = 0.1;

#[derive(Clone)]
pub struct Dirt {
    pub dampness: Arc<Mutex<f32>>,
    pub nutrients: Arc<Mutex<f32>>,
}

impl Dirt {
    pub fn new() -> Self {
        Self {
            dampness: Arc::new(Mutex::new(0.9)),
            nutrients: Arc::new(Mutex::new(1.0)),
        }
    }
}

impl Cell for Dirt {
    fn do_action(&self, position: Position) -> Option<Action> {
        let mut pending_changes: Vec<(Arc<dyn Cell>, f32)> = Vec::new();
        let mut current_dampness_change: f32 = 0.0;

        let initial_dampness = *self.dampness.lock().unwrap();

        if let Some(up) = grid::get_neighbor(position, Direction::Up) {
            if up.as_any().is::<Air>() && initial_dampness > MIN_DAMPNESS {
                current_dampness_change -= EVAPORATION_RATE;
            }
        }

        if let Some(down) = grid::get_neighbor(position, Direction::Down) {
            if let Some(dirt_down) = down.as_any().downcast_ref::<Dirt>() {
                let neighbor_dampness = *dirt_down.dampness.lock().unwrap();
                if initial_dampness > neighbor_dampness {
                    let amount_to_flow = (initial_dampness - neighbor_dampness) / 2.0 * GRAVITY_FLOW_RATE;
                    current_dampness_change -= amount_to_flow;
                    pending_changes.push((down.clone(), amount_to_flow));
                }
            }
        }

        if let Some(right) = grid::get_neighbor(position, Direction::Right) {
            if let Some(dirt_right) = right.as_any().downcast_ref::<Dirt>() {
                let neighbor_dampness = *dirt_right.dampness.lock().unwrap();
                if initial_dampness > neighbor_dampness {
                    let amount_to_flow = (initial_dampness - neighbor_dampness) / 2.0 * HORIZONTAL_FLOW_RATE;
                    current_dampness_change -= amount_to_flow;
                    pending_changes.push((right.clone(), amount_to_flow));
                }
            }
        }

        if let Some(left) = grid::get_neighbor(position, Direction::Left) {
            if let Some(dirt_left) = left.as_any().downcast_ref::<Dirt>() {
                let neighbor_dampness = *dirt_left.dampness.lock().unwrap();
                if initial_dampness < neighbor_dampness {
                    let amount_to_flow = (neighbor_dampness - initial_dampness) / 2.0 * HORIZONTAL_FLOW_RATE;
                    current_dampness_change += amount_to_flow;
                }
            }
        }

        if current_dampness_change.abs() > 0.0001 || !pending_changes.is_empty() {
            let self_dampness_arc = self.dampness.clone();

            return Some(Box::new(move || {
                let mut dampness = self_dampness_arc.lock().unwrap();
                *dampness += current_dampness_change;

                for (cell, dampness_delta) in pending_changes {
                    if let Some(target_dirt) = cell.as_any().downcast_ref::<Dirt>() {
                        let mut target_dampness = target_dirt.dampness.lock().unwrap();
                        *target_dampness += dampness_delta;
                    }
                }

                *dampness = dampness.clamp(MIN_DAMPNESS, MAX_DAMPNESS);
            }));
        }

        None
    }

    fn get_color(&self, _position: Position) -> String {
        let dry_color = (0xA0, 0x82, 0x61);
        let wet_color = (0x38, 0x25, 0x11);

        let t = (*self.dampness.lock().unwrap() / MAX_DAMPNESS).clamp(0.0, 1.0);

        let r = (dry_color.0 as f32 + (wet_color.0 as f32 - dry_color.0 as f32) * t) as u8;
        let g = (dry_color.1 as f32 + (wet_color.1 as f32 - dry_color.1 as f32) * t) as u8;
        let b = (dry_color.2 as f32 + (wet_color.2 as f32 - dry_color.2 as f32) * t) as u8;

        format!("#{:02X}{:02X}{:02X}", r, g, b)
    }
    fn clone_cell(&self) -> Arc<(dyn cells::Cell + 'static)> {
        let new_dirt = Dirt {
            dampness: Arc::new(Mutex::new(*self.dampness.lock().unwrap())),
            nutrients: Arc::new(Mutex::new(*self.nutrients.lock().unwrap())),
        };
        Arc::new(new_dirt)
    }
}